### DISCUSSION TITLE
Using OVER() in View Collections

### DISCUSSION THREAD

**Discussion by dimitarnikolovv** [pocketbase/pocketbase#3709] [Q&A] [answered]
I am trying to create a user rating system. I have a collection "ratings" containing the ratings and with relations to the user author and the targeted user.

I am trying to make a View collection that contains information about the rating per user basis, including the total ratings and the average.

That's what I came up with:

```sql
SELECT
    ru.username AS rater_username,
    ru.firstName AS rater_firstName,
    ru.lastName as rater_lastName,
    ru.id AS rater_id,
    r.rating AS user_rating,
    tu.id AS target_id,
    AVG(r.rating) OVER (PARTITION BY tu.id) AS averageRating,
    COUNT(*) OVER (PARTITION BY tu.id) AS totalRatings
FROM
    ratings r
JOIN
    users ru ON r.rater_id = ru.id
JOIN
    users tu ON r.target_id = tu.id;
```

However, I get "Invalid query - invalid identifier parts [AVG(r.rating) OVER (PARTITION BY tu.id) AS averageRating].", which really confuses me.

**Comment by ganigeorgiev** [comment]
I don't understand the above query nor why it is needed, but if an expression is not parsed properly you can try to wrap it in parenthesis, aka. `(...) AS yourFieldName`.

**Comment by dimitarnikolovv** [comment]
Basically, I want to have a collection storing all users' reviews. The reviews are user-to-user. I am trying to figure out how to separate the user-specific reviews in a View collection alongside the total reviews the user has received and the average number of all the user's reviews.

The reviews collection looks like this:
![Screenshot 2023-11-08 at 17 56 34](https://github.com/pocketbase/pocketbase/assets/49731322/b4575dec-fd75-4a62-a457-f1146e1802d6)

**Reply by ganigeorgiev** [reply, marked as answer]
I don't understand the difference between `rater_id` and `target_id` in the initial query. I guess that is the `author` and `target` fields in the above screenshot, right?

Wrapping the column expression in parenthesis `(...) as yourFieldName` should work.

I'm not sure whether you need the window function and the last join but I guess an alternative approach could be also using a simple `GROUP BY` statement. Here is a slightly modified version of your initial query:

```sql
SELECT
    r.taget as id,                 -- all view queries must have an id column
    ru.username AS rater_username,
    ru.firstName AS rater_firstName,
    ru.lastName as rater_lastName,
    ru.id AS rater_id,
    r.rating AS user_rating,
    r.taget,                       -- no need for join if you are just returning the field
    AVG(r.rating) AS averageRating,
    COUNT(*) AS totalRatings
FROM
    ratings r
JOIN
    users ru ON r.author = ru.id
GROUP BY r.target
```

**Reply by dimitarnikolovv** [reply]
Another question: If the user has a relation with the ratings collection: one-to-many, how I am supposed to expand it while getting the posts, where the user is a single relation?

**Reply by ganigeorgiev** [reply]
I'm sorry, I don't understand the question.

Multiple `relation` field values are stored as json array. If you wan to join each item from that relation, you can use the [SQLite json finctions](https://www.sqlite.org/json1.html). In your case I believe you need `json_each`:
```sql
...
left join json_each(yourRelField) as je
left join users on users.id = je.value
...
```

If that's not what you are looking for, please consider filing a new discussion with more details about your data structure and the expected outcome.
